### ISSUE TITLE
Polygon additions

### ISSUE THREAD

**Issue by hrishiballal** [Wildhoney/Leaflet.FreeDraw#63]
I am new to this library, I had a quick question regarding adding new polyons. I want to keep track of the Leaflet ID of every polygon. It seems like everytime a polygon is added new ids are assigned all the polygons, so there is no way to know which particular polygon is the "newest". 

I am not so much interested in the lat langs since I dont want to assign new ids everytime a polygon is added and I have set
```javascript
freeDraw.options.destroyPreviousPolygon(false);
freeDraw.options.allowMultiplePolygons(true);
```

On further investigation, here is what I found: 
1) Even after the options are set above, the 
```javascript
freeDraw.on('markers', function (e) {..
```
event is fired twice.  
2) After a new polygon is sketched, the library re-draws all the polygons by number of polygons +1 times.

**Comment by Wildhoney** [comment]
That's right, but it's not as simple as that because of polygon merging.

Assume you have 3 **separate** polygons on the map, you can certainly determine which is the new polygon by diffing which polygons you've seen before:

```javascript
const seenPolygons = [];

freeDraw.on('markers', function getMarkers(eventData) {

    var latestPolygon = eventData.latLngs.filter(function(latLngs) {
        const json = JSON.stringify(latLngs);
        const has  = !~seenPolygons.indexOf(json);
        (has) && seenPolygons.push(json);
        return has;
    });

    console.log('Added Polygon:', latestPolygon);

});
```

However, if you draw a polygon that encompasses all three of those polygon, and therefore merging 4 polygons together, those previous polygons are now invalid, and the encompassing polygon is the one you need. You'll therefore also need to also add some logic to detect which polygons were *removed*.

**Comment by hrishiballal** [comment]
Thank you for your comment, agreed this till give me the added polygon, however, in addition, this means that there needs to be a logic for edited as well. What I trying to do is the following: 
1) Enable the user to sketch two types of polygons: one red and one green, at any time they can edit / add / remove polygons. 
2) I start with drawing the red polygon, after it is complete, I select drawing the green polygon 
3) after finishing the drawing of the green polygon, since the red one (previously drawn) is recreated, how do I paint it back to red? If the ID was not recreated, I could track the color by ID but it is not possible. 
2) I am happy to turn off polygon merges

**Comment by hrishiballal** [comment]
Your code above will work for add / remove but will not work for edits.

**Comment by jekuno** [comment]
@hrishiballal What you are trying to do (drawing polygons in different colors) sounds similiar to my plans described in #53. My approach was to create a separate freedraw instance per color.

**Comment by hrishiballal** [comment]
Yes it seems that is the way to implement it, but I can see it being a problem for a lot of colors > 20. Also does this mean that I will need two edit / delete buttons ( one for every layer)?

**Comment by jekuno** [comment]
You don't need separate edit/delete buttons. If a user clicks on the "red color polygon tool" you can activate e.g. the CREATE, EDIT and DELETE mode for this freeDraw instance so that the user can either draw new red polygons or edit existing red polygons or delete existing red polygons by clicking into them.
If you have more than 20 colors you probably wouldn't add 20 buttons but make some dropdown to select the desired color.

In my discussion with @Wildhoney at #53 we agreed that it would be nice to have only one FreeDraw instance which supports different polygon types/colors so that you can e.g. invoke `setCurrentType('water')` to set the type to 'water' which then uses the blue css classes for icon and polygon to paint blue colored polygons.

**Comment by hrishiballal** [comment]
Yes that would be really great especially since it is not just about color, it other metadata and properties that can be attached to a drawn element (that are not lost).

**Comment by hrishiballal** [comment]
How to enable click to edit / delete? I am manually setting the modes currently.

**Comment by jekuno** [comment]
When you set the delete mode you can click into a polygon to delete it. See http://freedraw.herokuapp.com/

**Comment by hrishiballal** [comment]
Yes but how to capture what polygon has been clicked on? 
```
 If a user clicks on the "red color polygon tool" you can activate e.g. the CREATE, EDIT and DELETE mode for this freeDraw instance....
````

**Comment by jekuno** [comment]
I am talking about a click on the tool (i.e. the button for switching to the according color) not about a click on the polygon itself.

**Comment by hrishiballal** [comment]
Ah got it! So enable edit, delete and create mode all in one for that freedraw instance when the . OK understood.

**Comment by hrishiballal** [comment]
I have been implementing this solution now and have a couple of issues: 
I have added two freedraw instances and bound them on the same map. My user interface I have two buttons, one for red and other for green and enabling them sets EDIT, DELETE and CREATE (EDC) modes for that specific freedraw instance (and turns the other to "View" mode). 

I can create polgyons with this no problems. However, when I press the EDC mode for green layer, the elbows of the red layer are also activated. I tried to set the icon class name for the second layer:

```javascript
freeDrawG.options.setIconClassName('edge-green');
freeDrawG.options.setPolygonClassName('polygon-green');
```

How come elbows are activated for the second layer when I am activating it for the first one (or vice versa)?

**Comment by jekuno** [comment]
Might this be related to the issue I filed in #53 as reproduced in http://jsfiddle.net/1smuagqz/2/?

**Comment by hrishiballal** [comment]
Yes it is similar. I cannot share the entire code, since it is integrated into a very large application following are some snippets: 
1) Radio buttons to switch views: 
![controls](https://cloud.githubusercontent.com/assets/1588744/10453689/4483fa16-71a6-11e5-80dc-16c999f56bde.png)
2) Add two freedraw layers to the map
```javascript
var freeDraw = new L.FreeDraw({
    mode: L.FreeDraw.MODES.ALL
  });
  freeDraw.options.destroyPreviousPolygon(false);
  freeDraw.options.allowMultiplePolygons(true);
  freeDraw.options.exitModeAfterCreate(true);
  freeDraw.on('markers', function getMarkers(e) {
    var evt = e.target;
    if (evt.polygonCount >0)
    {
        var sty = getFreeDrawStyle('red');
         for (var k = 0, pcount = evt.polygons.length; k < pcount; k++) {
           var p = evt.polygons[k];
           p.setStyle(sty);
         }
         }
    });

  freeDraw.setMode(L.FreeDraw.MODES.VIEW);

  mapDetails.freeDrawFeat = freeDraw;
  map.addLayer(freeDraw);

    // create a new freedraw for green. 
  var freeDrawG = new L.FreeDraw({
    mode: L.FreeDraw.MODES.ALL
  });
  freeDrawG.options.destroyPreviousPolygon(false);
  freeDrawG.options.allowMultiplePolygons(true);
  freeDrawG.options.exitModeAfterCreate(true);

  freeDrawG.on('markers', function getMarkers(eg) {
    var evtg = eg.target;
    
         if (evtg.polygonCount >0)
         {
          var sty = getFreeDrawStyle('green');
          for (var j = 0, pcountg = evtg.polygons.length; j < pcountg; j++) {
           var p = evtg.polygons[j];
           p.setStyle(sty);
         }
       }


  });
  mapDetails.freeDrawFeatG = freeDrawG;
  map.addLayer(freeDrawG);
```
2) in the 'onchange' of radio button
```javascript
if (mode === 'fdcreatered') {
  FreedrawColor = 'red';
  SelectedAreaType = 'red';
  freeDraw.options.polygonType = 'red';
  freeDrawG.setMode(L.FreeDraw.MODES.VIEW);
  freeDraw.setMode(L.FreeDraw.MODES.CREATE | L.FreeDraw.MODES.EDIT | L.FreeDraw.MODES.DELETE); 
  
} else if (mode === 'fdcreategreen') {
  FreedrawColor = 'green';
  SelectedAreaType = 'green';
  freeDrawG.options.polygonType = 'green';
  freeDraw.setMode(L.FreeDraw.MODES.VIEW);
  freeDrawG.setMode(L.FreeDraw.MODES.CREATE | L.FreeDraw.MODES.EDIT | L.FreeDraw.MODES.DELETE);

} else if (mode === 'fdview') {
      freeDraw.setMode(L.FreeDraw.MODES.VIEW);
      freeDrawG.setMode(L.FreeDraw.MODES.VIEW);
}
```
Flow: 
Step 1) Create red, draw a polygon
![create-red1](https://cloud.githubusercontent.com/assets/1588744/10453837/947e79dc-71a7-11e5-88f5-91c25dedfdd3.png)

Step 2) Click on View button (set both layers to view mode)
![create-red-view](https://cloud.githubusercontent.com/assets/1588744/10453842/a055f032-71a7-11e5-9381-2a6097750600.png)

Step 3) Click on Create green button. ERROR, notice how existing Red elbows are selected. 
![create-green-error](https://cloud.githubusercontent.com/assets/1588744/10453853/b356c5c6-71a7-11e5-9c00-39f85eeb5f6a.png)
Step 4) Click on View and again click on Create Red. ERROR: the green elbows are selected. 
![create-green-error](https://cloud.githubusercontent.com/assets/1588744/10453864/c78f972a-71a7-11e5-94c0-69a7fc18f26c.png)

**Comment by jekuno** [comment]
Thanks for the detailed documentation of the issue.

As I understand what @Wildhoney says in https://github.com/Wildhoney/Leaflet.FreeDraw/issues/53#issuecomment-98765806 there are problems with multiple `FreeDraw` layers because `Leaflet` binds events on the map and not on separate layers.

That's why you, @hrishiballal, probably also need to try to work with only ONE `FreeDraw` layer.
To make this working @Wildhoney started to work on #53 and committed some event fixes in https://github.com/Wildhoney/Leaflet.FreeDraw/commit/65b23c5fb20536111b24bb126b4e1f3619f236af. This commit however only supports one polygon per color which doesn't suffice.

This is how Leaflet.FreeDraw currently determines the elbow color:
https://github.com/Wildhoney/Leaflet.FreeDraw/blob/65b23c5fb20536111b24bb126b4e1f3619f236af/components/FreeDraw.js#L1249
We need some logic to determine the correct color for each polygon there. Best would be if we wouldn't operate on plain `L.polygon` instances here but if we would get an `L.FreeDraw.Polygon` instance with an additional property such as `type` (which can e.g. have values `water` and `land`).

If the polygon type e.g. is `water`, then `Leaflet.FreeDraw` would simply set the `iconClassName` for this polygon to `iconClassWater` and the `polygonClassName` for this polygon to `polygonClassWater`. Then everything would be fine and the polygon could be drawn with e.g. blue fill color and blue elbows.

**Comment by hrishiballal** [comment]
I see, this is a bit unfortunate. What is the way out of this? My application needs to have different polygon colors. I am sure this is quite common. 

@Wildhoney I am happy to do a very modest fee contribution to the library for this if it can be prioritized.

**Comment by Wildhoney** [comment]
Two polygons of different colours certainly wasn't in the initial plan for this module. If you folks want to try and add it to the module, then pull requests are more than welcome, considering you don't break existing functionality or complicate the interface.

**Comment by hrishiballal** [comment]
Can you point to a place to start please? I am not in a position to contribute because of my schedule currently but I can start looking at it.

**Comment by Wildhoney** [comment]
Well as @jekuno pointed out, I've already implemented part of it with https://github.com/Wildhoney/Leaflet.FreeDraw/issues/53 &ndash; it just didn't go far enough to what he wanted. I'd definitely use that as a starting point.

I wouldn't recommend going down the two instance route though, as that seems like a hack.

**Comment by hrishiballal** [comment]
I looked at the code and tried a couple of approaches and I do not think that it will work. Here is what I tried: Store all polygons from getPolygons in a object with their type and try to link them to createPolygon method , however because the Clipper library is being used to recreate the polygons, the state information is lost. Furthermore the code recreates all polygons on the layer and not just the one edited / added so there is really no way to track changes on a polygon. 
@jekuno, the issue with your idea of polygon type is that it is no different than classname, infact i tried to put a dictionary instead of a string for classname, because the layers are recreated everytime, there is no way to "get/set" types since the recreation means that the old features are destoryed (or merged) 

The way out seems to be two things: maintaining state information during createPolygons and then modes (e.g. EDIT, DELETE etc.) tied to class name so features of the same class are selected for edit, delete (but not of the other class). I am not able to do a PR for this.

**Comment by hrishiballal** [comment]
I was able to implement a two polygon drawing code using the steps mentioned earlier in this thread, only editing is disabled and also merging. Once we have that, the polygons can be tracked. I use _originalPoints to check the polygons since that seems to the best way to be sure. Closing this bug.

**Comment by jekuno** [comment]
Great you achieved some progress here. Let me ask you some questions:

1. Do you find the matching polygon by iterating over the existing polygons and comparing all points?
2. Would you propose changes in `Leaflet.FreeDraw` to make it easier to identify the matching polygon?
3. Do you plan to support editing as well?
4. Could you share the code so that the community can maybe cowork on this?

**Comment by hrishiballal** [comment]
Hi @jekuno , sorry for the late reply, I have been busy testing the implementation. 

1) Yes however I use _originalpoints array and check that (length and content) 
2) I looked at polygon hashing but for the moment just checking the length of originalpoints and latlng contents seems to work. I am not sure how to hash geojson polygons, another angle I looked at was polygon area using GeoJSON utils. 
3) No at the moment editing cannot be supported because once it is turned on there is no way to know what is already there and what was deleted. In the situation where more than one color is present, editing cannot be permitted (only add or remove). 
4) The code is actually quite simple, just track what the user has clicked (edit / create red or green) and based on that store the red and green polygons in different arrays and check _originalpoints. 

I hope it helps. I am using it only for two colors, but for more than that I think that it will become very difficult. I have used Leaflet.draw for a number of years and the geojson native implementation is great but this library provides a more natural drawing experience, I wish it supported GeoJSON better but oh well. 

PS I mix L.Pather along with this one on the same map so people can draw either line or polygons (they select first which one)  and the appropriate layer is loaded.
